Additional fixes and improvements to JavaClassWrapper

m4gr3d opened the pull request:
Follow up to https://github.com/godotengine/godot/pull/96182

- Fix crashing bug when invoking Java class constructor with parameters
- Add support for accessing class constants
- Add support for Godot `Callable` arguments. A Godot `Callable` can be wrapped by a Java `Runnable` to allow Java logic to run arbitrary Godot lambdas
- Automatically convert `java.lang.CharSequence` to Godot `String` as needed
- Code cleanup

Those fixes make https://github.com/godotengine/godot-proposals/issues/11189 feasible out of the box.

<!--
Please target the `master` branch in priority.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.

To speed up the contribution process and avoid CI errors, please set up pre-commit hooks locally:
https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html
-->

m4gr3d committed:
Additional fixes and improvements to JavaClassWrapper

- Fix crashing bug when invoking class constructor with parameters
- Add support for accessing class constants
- Add support for Godot Callable arguments. A Godot Callable can be wrapped by a Java Runnable to allow Java logic to run arbitrary Godot lambdas
- Automatically convert java.lang.CharSequence to Godot String as needed
- Code cleanup

syntaxerror247 commented:
Hi @m4gr3d The code looks good! However, I’m having trouble with the toast message example you provided—it’s not working on my end. I’m not seeing any error messages but it looks like the callable is never called.

I have tested it on Motorola (android 11) and Xiaomi (android 14).

dsnopek reviewed (commented):
The high-level idea of this looks great!

However, I suspect that some of the memory management around callables (particularly the `Variant *` that it hangs on to) isn't quite right. More details in the code comments.

I haven't had a chance to properly test it yet, though.

m4gr3d commented on platform/android/jni_utils.cpp:
```diff
@@ -32,6 +32,56 @@
+
+	jclass bclass = p_env->FindClass("org/godotengine/godot/variant/Callable");
+	jmethodID ctor = p_env->GetMethodID(bclass, "<init>", "(J)V");
+	jobject jcallable = p_env->NewObject(bclass, ctor, reinterpret_cast<int64_t>(p_callable));
```
@dsnopek Thanks for the suggestion! I've updated the PR, and that also fixes the issue @syntaxerror247 was seeing where the toast doesn't show again. @syntaxerror247 Can you test again and check if the callable is called and the toast shows?

> and having the Callable class memdelete() it up when it is destroyed.

This part is more complex to do because java objects don't get callbacks when they're destroyed, so we don't have a mechanism to know when to clean up. I've omitted it for now, which means we're allocating `Variant *` but not freeing, so looking for suggestions.

syntaxerror247 replied:
@m4gr3d I tested again, now it works!

m4gr3d replied:
> This part is more complex to do because java objects don't get callbacks when they're destroyed, so we don't have a mechanism to know when to clean up. I've omitted it for now, which means we're allocating Variant * but not freeing, so looking for suggestions.

I ended up using the [Object::finalize()](https://developer.android.com/reference/java/lang/Object#finalize()) method in order to detect when to deallocate the native memory. 
As marked in the documentation, that method has been deprecated; unfortunately its replacement, [`java.lang.ref.Cleaner`](https://developer.android.com/reference/java/lang/ref/Cleaner) is only supported on Android api 33 and higher, so we'll need to update that logic when our min api target catches up.

dsnopek replied:
The solution to this issue in the latest version of the PR looks good to me!

dsnopek commented on platform/android/variant/callable_jni.cpp:
```diff
@@ -0,0 +1,117 @@
+
+	int count = p_env->GetArrayLength(p_parameters);
+
+	Variant *args = (Variant *)alloca(sizeof(Variant) * count);
+	const Variant **argptrs = (const Variant **)alloca(sizeof(Variant *) * count);
+
+	for (int i = 0; i < count; i++) {
+		jobject jobj = p_env->GetObjectArrayElement(p_parameters, i);
+		ERR_FAIL_NULL_V(jobj, Callable());
+		memnew_placement(&args[i], Variant(_jobject_to_variant(p_env, jobj)));
+		argptrs[i] = &args[i];
+		p_env->DeleteLocalRef(jobj);
+	}
+
+	return Callable(obj, str_method).bindp(argptrs, count);
```
I think this will lead to a memory leak, but I'm not 100% sure.

The reason I suspect it is that I don't think the destructors for each `Variant` in `args` will run automatically. This is fine for any variant type where all the data is part of the `Variant` class (because this uses `alloca()` to put their data on the stack), but some types refer to memory elsewhere (for example: `String`s, `RefCounted` object pointers or any of the `Packed*Array` types) and use reference counting, which won't be decremented when this function exits, leaving the ref count artificially inflated by one.

So, I think we need to add some code at the end that manually calls `~Variant()`, like:

```suggestion
	Variant *args = (Variant *)alloca(sizeof(Variant) * count);
	const Variant **argptrs = (const Variant **)alloca(sizeof(Variant *) * count);

	for (int i = 0; i < count; i++) {
		jobject jobj = p_env->GetObjectArrayElement(p_parameters, i);
		ERR_FAIL_NULL_V(jobj, Callable());
		memnew_placement(&args[i], Variant(_jobject_to_variant(p_env, jobj)));
		argptrs[i] = &args[i];
		p_env->DeleteLocalRef(jobj);
	}

	Callable ret = Callable(obj, str_method).bindp(argptrs, count);
	
	for (int i = 0; i < count; i++) {
		args[i].~Variant();
	}
	
	return ret;
```

I don't think this should cause any problems, because `.bindp()` will copy each `Variant`, which will increment the ref count (if there is a refcount for that particular variant type).

dsnopek commented on platform/android/variant/callable_jni.cpp:
```diff
@@ -0,0 +1,117 @@
+
+	int count = p_env->GetArrayLength(p_parameters);
+
+	Variant *args = (Variant *)alloca(sizeof(Variant) * count);
+	const Variant **argptrs = (const Variant **)alloca(sizeof(Variant *) * count);
+
+	for (int i = 0; i < count; i++) {
+		jobject jobj = p_env->GetObjectArrayElement(p_parameters, i);
+		ERR_FAIL_NULL_V(jobj, nullptr);
+		memnew_placement(&args[i], Variant(_jobject_to_variant(p_env, jobj)));
+		argptrs[i] = &args[i];
+		p_env->DeleteLocalRef(jobj);
+	}
+
+	Callable callable = *callable_variant;
+	if (callable.is_valid()) {
+		Callable::CallError err;
+		Variant result;
+		callable.callp(argptrs, count, result, err);
+		jvalret jresult = _variant_to_jvalue(p_env, result.get_type(), &result, true);
+		return jresult.obj;
+	} else {
+		return nullptr;
+	}
```
I think this has the same problem...

dsnopek replied:
It looks like `Java_org_godotengine_godot_plugin_GodotPlugin_nativeEmitSignal()` (which isn't modified in this PR) also shares this same code

dsnopek reviewed (commented):
Thanks!

The latest code changes look good to me as far as fixing the memory management issue that I commented on in my last review! However, while looking at the PR again with memory management on my mind, I think I may have found another issue - a memory leak this time. See the code comments below for more details.

I also did some quick testing: basically, just trying the example given in the code comments in this PR. It worked great!

llama-nl commented:
I have a question:
based on this https://developer.android.com/reference/android/os/Build

I was trying this code:
```gdscript
var android_runtime = Engine.get_singleton("AndroidRuntime")
if android_runtime:
	var build_class = JavaClassWrapper.wrap("android.os.Build") #JavaCalss  <android.os.Build>
	var model = build_class.BRAND
	print("Device Brand: %s (%s)" % [model])
else:
	printerr("Couldn't find AndroidRuntime singleton.")
```
And it was showing this error:
```
Invalid access to property or key 'BRAND' on a base object of type 'JavaClass'.

Error
```

m4gr3d commented:
@dsnopek I've addressed your feedback!

@llama-nl I've also fixed access to the class constants. Your code sample should now work with the latest version of this PR.

dsnopek reviewed (approved):
Thanks! This looks great to me :-)

> I've also fixed access to the class constants.

Oh, awesome! I hadn't found the time to dig into it myself yet

akien-mga commented:
Thanks!

SakuyaCN commented:
If my parameters need to pass an interface as a callback, it seems like it wouldn’t be feasible to implement it directly. I would need to implement the interface at the Java layer and then use JavaClassWrapper to create a class to serve as the parameter.
like this:
`public interface Callback<RESULT> {
    public abstract fun onFail(): kotlin.Unit

    public abstract fun onSuccess(result: RESULT): kotlin.Unit
}


AnyObject.useCallback(object:Callback<RESULT>{ ... });
`

SakuyaCN commented:
If my parameters need to pass an interface as a callback, it seems like it wouldn’t be feasible to implement it directly. I would need to implement the interface at the Java layer and then use JavaClassWrapper to create a class to serve as the parameter.
like this:
```
public interface Callback<RESULT> {

    public abstract fun onFail(): kotlin.Unit

    public abstract fun onSuccess(result: RESULT): kotlin.Unit

}
```


`AnyObject.useCallback(object:Callback<RESULT>{ ... });`

SakuyaCN commented:
If the method being called has a parameter that is a callback, and this callback is a custom interface, do I need to implement this interface at the Java level and then use JavaClassWrapper?

dsnopek commented:
@SakuyaCN To implement an interface, you'd have to do it on the Java side. `JavaClassWrapper` only lets you call methods or create instances of already existing classes, not define new ones.

And if you're going to some stuff on the Java side (with a [Godot Android plugin](https://docs.godotengine.org/en/latest/tutorials/platform/android/android_plugin.html)) then, depending on your needs, you may want to do everything that way and not use `JavaClassWrapper` at all. The primary goal of `JavaClassWrapper` is to allow you to do simple things from GDScript/C# without having to switch to the Java side.

m4gr3d commented:
> If the method being called has a parameter that is a callback, and this callback is a custom interface, do I need to implement this interface at the Java level and then use JavaClassWrapper?

@SakuyaCN Can you provide some sample code (pseudo-code is fine) to illustrate what you are asking?

SakuyaCN commented:
Another issue is that if a class has two overloaded methods, as shown in the example below, using call('<init>') always invokes the first one.
```
    TestClass(String name,String[] args)
    
    TestClass(String name,Object obj)
```

dsnopek commented:
> There's another problem. Suppose a class has two overloaded methods, as demonstrated below. When using` call('<init>')` and passing parameters, regardless of the type of the second parameter I pass, the first method is always called.

This one sounds like a bug. Can you open a new issue for it?

FYI, instead of `call('<init>', ...)` you can use the class name, so: `TestClass.TestClass(...)`
